injector/injector.go

```diff
@@ -0,0 +1,343 @@
+	return &TrackerInjector{
+		replacer: replacer,
+		provider: provider,
+		events:   events,
+	}
+}
+
+func (ti *TrackerInjector) InjectTracker(vastXML string, NURL string) string {
+	if vastXML == "" && NURL == "" {
+		// TODO Log a adapter.<bidder-name>.requests.badserverresponse
+		return vastXML
+	}
+
+	if vastXML == "" {
+		return fmt.Sprintf(emptyAdmResponse, NURL)
+	}
+
+	var outputXML strings.Builder
+	encoder := xml.NewEncoder(&outputXML)
+	st := &InjectionState{
+		injectTracker:         false,
+		injectVideoClicks:     false,
+		inlineWrapperTagFound: false,
+		wrapperTagFound:       false,
+		impressionTagFound:    false,
+		errorTagFound:         false,
+		creativeId:            "",
+		isCreative:            false,
+		companionTagFound:     false,
+		nonLinearTagFound:     false,
+	}
+
+	reader := strings.NewReader(vastXML)
+	decoder := xml.NewDecoder(reader)
+
+	for {
+		t, err := decoder.RawToken()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return ""
+		}
+
+		switch tt := t.(type) {
+		case xml.StartElement:
+			ti.handleStartElement(tt, st, &outputXML, encoder)
+		case xml.EndElement:
+			ti.handleEndElement(tt, st, &outputXML, encoder)
+		case xml.CharData:
+			tt2 := strings.Trim(string(tt), trimRunes)
+			if len(tt2) != 0 {
+				encoder.Flush()
+				outputXML.WriteString("<![CDATA[")
+				outputXML.WriteString(tt2)
+				outputXML.WriteString("]]>")
+				continue
+			}
+		default:
+			encoder.EncodeToken(t)
+		}
+	}
+
+	encoder.Flush()
+
+	if !st.inlineWrapperTagFound {
+		// Todo log adapter.<bidder-name>.requests.badserverresponse metrics
+		return vastXML
+	}
+	return outputXML.String()
+}
+
+func (ti *TrackerInjector) handleStartElement(tt xml.StartElement, st *InjectionState, outputXML *strings.Builder, encoder *xml.Encoder) {
+	switch tt.Name.Local {
+	case wrapperCase:
+		st.wrapperTagFound = true
+		encoder.EncodeToken(tt)
+	case creativeCase:
+		st.isCreative = true
+		for _, attr := range tt.Attr {
+			if strings.ToLower(attr.Name.Local) == adId {
+				st.creativeId = attr.Value
+			}
+		}
+		encoder.EncodeToken(tt)
+	case linearCase:
+		st.injectVideoClicks = true
+		st.injectTracker = true
+		encoder.EncodeToken(tt)
+	case videoClicksCase:
+		st.injectVideoClicks = false
+		encoder.Flush()
+		encoder.EncodeToken(tt)
+		encoder.Flush()
+		ti.addClickTrackingEvent(outputXML, st.creativeId, false)
+	case nonLinearAdsCase:
+		st.injectTracker = true
+		encoder.EncodeToken(tt)
+	case trackingEventsCase:
+		if st.isCreative {
+			st.injectTracker = false
+			encoder.Flush()
+			encoder.EncodeToken(tt)
+			encoder.Flush()
+			ti.addTrackingEvent(outputXML, st.creativeId, false)
+		}
+	default:
+		encoder.EncodeToken(tt)
+	}
+}
+
+func (ti *TrackerInjector) handleEndElement(tt xml.EndElement, st *InjectionState, outputXML *strings.Builder, encoder *xml.Encoder) {
+	switch tt.Name.Local {
+	case impressionCase:
+		encoder.Flush()
+		encoder.EncodeToken(tt)
```

Comment:
Same comment about `EncodeToken` possibly being moved to the end of this function.

Reply:
I don't know if you forgot to post the reasoning, but why can't `EncodeToken` be moved?